test/gpu/prefuse_ops.cpp

```diff
@@ -0,0 +1,204 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <migraphx/dead_code_elimination.hpp>
+#include <migraphx/gpu/prefuse_ops.hpp>
+#include <migraphx/gpu/fuse_mlir.hpp>
+#include <migraphx/gpu/gemm_softmax_gemm.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/generate.hpp>
+#include <basic_ops.hpp>
+#include <migraphx/make_op.hpp>
+#include <test.hpp>
+
+struct pre_gemm_softmax_gemm : migraphx::gpu::gemm_softmax_gemm
+{
+    std::string name() const { return "gpu::pre_gemm_softmax_gemm"; }
+};
+
+void run_pass(migraphx::program& p)
+{
+    migraphx::run_passes(p, {migraphx::gpu::prefuse_ops{}, migraphx::dead_code_elimination{}});
+}
+
+TEST_CASE(find_gemm_softmax_gemm)
+{
+    migraphx::shape s1{migraphx::shape::float_type, {8, 16, 32}};
+    migraphx::shape s2{migraphx::shape::float_type, {8, 32, 16}};
+
+    auto create_program = [=]() {
+        migraphx::program p;
+        auto* mm   = p.get_main_module();
+        auto x     = mm->add_parameter("x", s1);
+        auto y     = mm->add_parameter("y", s2);
+        auto z     = mm->add_parameter("z", s1);
+        auto scale = mm->add_literal(2.0f);
+
+        auto dot1     = mm->add_instruction(migraphx::make_op("dot"), x, y);
+        auto scale_mb = mm->add_instruction(
+            migraphx::make_op("multibroadcast", {{"out_lens", dot1->get_shape().lens()}}), scale);
+        auto mul  = mm->add_instruction(migraphx::make_op("mul"), dot1, scale_mb);
+        auto sm   = mm->add_instruction(migraphx::make_op("softmax", {{"axis", 2}}), mul);
+        auto dot2 = mm->add_instruction(migraphx::make_op("dot"), sm, z);
+        mm->add_return({dot2});
+        return p;
+    };
+
+    auto create_fused_program = [=]() {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+        auto x   = mm->add_parameter("x", s1);
+        auto y   = mm->add_parameter("y", s2);
+        auto z   = mm->add_parameter("z", s1);
+
+        auto attn =
+            mm->add_instruction(pre_gemm_softmax_gemm{migraphx::make_op("dot"), 2}, x, y, z);
+
+        mm->add_return({attn});
+        return p;
+    };
+
+    migraphx::program p1 = create_program();
+    migraphx::program p2;
+    if(migraphx::gpu::mlir_attention_enabled())
```

Comment:
Dont rely on this global setting. Instead add a flag to the `prefuse_ops` class like `enable_attention` that can be set to true during tests. And then you pass that along to the `find_gemm_softmax_gemm` so the matcher will match it.